### ISSUE TITLE
XML plists opened as Old-Style

### ISSUE THREAD

**Issue by gfpacheco** [atom/language-property-list#5]
I remember working on plist files in atom, the syntax highlight worked back there. But after not opening a plist file for a long time they now look like this:

![screen shot 2016-02-11 at 11 00 29](https://cloud.githubusercontent.com/assets/3705660/12976982/f8efd8f0-d0ae-11e5-9659-7b9d8a0496f6.png)

This is the content of the file:

``` plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>CFBundleDevelopmentRegion</key>
    <string>en</string>
    <key>CFBundleExecutable</key>
    <string>$(EXECUTABLE_NAME)</string>
    <key>CFBundleIdentifier</key>
    <string>$(PRODUCT_BUNDLE_IDENTIFIER)</string>
    <key>CFBundleInfoDictionaryVersion</key>
    <string>6.0</string>
    <key>CFBundleName</key>
    <string>$(PRODUCT_NAME)</string>
    <key>CFBundlePackageType</key>
    <string>APPL</string>
    <key>CFBundleShortVersionString</key>
    <string>1.0</string>
    <key>CFBundleSignature</key>
    <string>????</string>
    <key>CFBundleVersion</key>
    <string>1</string>
    <key>LSRequiresIPhoneOS</key>
    <true/>
    <key>UILaunchStoryboardName</key>
    <string>LaunchScreen</string>
    <key>UIRequiredDeviceCapabilities</key>
    <array>
        <string>armv7</string>
    </array>
    <key>UISupportedInterfaceOrientations</key>
    <array>
        <string>UIInterfaceOrientationPortrait</string>
    </array>
</dict>
</plist>
```

Does it work for you guys?

**Comment by lee-dohm** [comment]
I can't reproduce what you're seeing using Atom v1.7.0-dev-d77c334 on Mac OS X 10.11.3:

<img width="928" alt="screen shot 2016-02-11 at 11 55 24 am" src="https://cloud.githubusercontent.com/assets/1038121/12988641/6fd7a3f6-d0b6-11e5-81fb-a1d55f9ad261.png">

Do you have the "Property List (XML)" grammar selected?

<img width="217" alt="screen shot 2016-02-11 at 11 55 33 am" src="https://cloud.githubusercontent.com/assets/1038121/12988647/7d2906b2-d0b6-11e5-8cc5-3254a6412cf5.png">

**Comment by gfpacheco** [comment]
There are two grammars inside this package:

- `Property List (Old-Style)` that is scoped to `source.plist`
- `Property List (XML)` that is scoped to `source.xml.plist`

If I manually select the XML one it works fine, but the scopes tell me that it won't be selected unless the file name ends with `.xml.plist`. Did I understand it right?

**Comment by lee-dohm** [comment]
Actually, it will select the XML one based on the "first line" of the file:

https://github.com/atom/language-property-list/blob/master/grammars/property%20list%20(xml).cson#L2

I just used "Auto Detect" on my untitled buffer and it detected as "Property List (XML)" just fine. If I save it as `test.plist` I see the behavior you're seeing because it switches to "Property List (Old-Style)" based on the extension.

**Comment by lee-dohm** [comment]
If you add this to the `core` section in your `config.cson`:

```
    customFileTypes:
      'text.xml.plist': ['plist']
```

Atom will default to treating files with the `plist` extension as "Property List (XML)" files.

**Comment by gfpacheco** [comment]
Great! Thank you a lot.

Can this be fixed to work out of the box, or it'll still need this workaround?

**Comment by 50Wliu** [comment]
The firstLineMatch should probably be given higher priority than the file type.

**Comment by lee-dohm** [comment]
@50Wliu I think that would work well in this case. I'm not sure if that would be the best for all cases though. Changing to prefer the first line match rules over file type might have performance implications too.

**Comment by lee-dohm** [comment]
> Can this be fixed to work out of the box, or it'll still need this workaround?

A change to make this the default behavior, i.e. treat files with extension `plist` as "Property List (XML)", is possible. Whether that is the best solution is unclear.

**Comment by Redoubts** [comment]
Old style plists are really old, and the documentation says you can't easily create them anymore 
https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/PropertyLists/OldStylePlists/OldStylePLists.html

> These files store information equivalent to the information in XML property lists, and are still supported by Cocoa, but for reading only. Old-style plist support remains primarily for legacy reasons.

I think making XML default for .plist makes a lot of sense, I'm not sure why you'd want the next-step version to be default. I don't think I've even seen them in the wild.

**Comment by justinpawela** [comment]
I can confirm that I get "Property List (Old Style)" every time as well, instead of the expected XML variant.

OS X 10.11.4, Atom 1.6.1

I'm really pretty sure this has not always been the case. Unfortunately, I don't know at what point the change occurred. However, it doesn't look like much as changed in the codebase here in quite some time, especially in the grammar definitions where the file types and first line patterns are defined. It's been more than two years since those changed.

I am not familiar with the Atom codebase, but is there something in the app that could have changed to affect file type resolution? I think within the last 3 months; certainly within the last 6 months.

Just opened an `Info.plist` file and got the "Old Style" (first few lines below):

```plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
```

**Comment by lee-dohm** [comment]
@justinpawela No, I can't think of anything that would have changed in the last six months that would have changed this behavior. You could always install an older version and see?

**Comment by Ingramz** [comment]
I noticed this too and realized that I had broken it with https://github.com/atom/atom/pull/9513 - oops! 😨 

But I believe I also have a solution. If we add the possible file types to the XML grammar, the grammar scoring system should favor the XML in the case where the first line of the file matches the regular expression specified in that grammar.

If neither of the first line regular expressions matches, we have a slight problem as the first grammar that is encountered, will be chosen, unfortunately I don't know what the ordering is based on (and whether we could rely on it), but right now the old style grammar is listed before the XML version, which hopefully is what we want to fall back to.

So in conclusion I believe, defining file types for XML grammar should fix the issue we are having here.

**Comment by brettimus** [comment]
Also been experiencing this! Feelin' the p-list pain

Atom 1.7.3, OS X 10.11.5

**Comment by pulimento** [comment]
Atom 1.7.3, OS X 10.11.5, can confirm that @lee-dohm fix [(see here)](https://github.com/atom/language-property-list/issues/5#issuecomment-183048068) does the trick

**Comment by brettimus** [comment]
@pulimento thanks!

**Comment by d4nirod** [comment]
Atom 1.7.4, OSX 10.11.5: @lee-dohm fix is not working for me. Still detecting the .plist file as `Property List (Old Style)`:
this is an excerpt from my modified `config.cson`:
```"*":
  core:
    customFileTypes:
      "text.xml.plist": ["plist"
    ]
    disabledPackages: [
      "language-gfm"
      "pretty-json"
    ]
...
```
And this is the beggining of the offending file:
```<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>```
Any ideas to make this work without having to manually change the grammar to `Property List (XML)`.

**Comment by d4nirod** [comment]
Atom 1.7.4, OS X 10.11.5: trick works here also.

**Comment by gingerbeardman** [comment]
Thanks for the workaround

**Comment by umeditor** [comment]
Still an issue with Atom 1.9.9 on 10.11.6. config.cson workaround does the trick.

**Comment by Ingramz** [comment]
@lee-dohm is there anything that's stopping #6 from being merged and solving this?

**Comment by gdombiak** [comment]
Atom 1.10.2 + MacOS Sierra = same problem.

**Comment by paustin01** [comment]
+1

**Comment by kyoh86** [comment]
+1

Atom 1.11.2, OS X 10.12.1

**Comment by ianpogi5** [comment]
Issue has been there for as long as I remember (maybe a year). I've fixed with @lee-dohm solution.

Atom 1.12.2, OS X 10.12.1

**Comment by gingerbeardman** [comment]
Why was this closed? Is it finally fixed?

**Comment by skipperen** [comment]
Not sure what was fixed, but the issue was still there initially, tested with Atom 1.12.5 on a Mac running 10.10.5 and on one running 10.12.1.  

After using the  @lee-dohm workaround, the plist files now open with XML grammar as desired, and pasted plist in a new document show as XML grammar as well.  

The Atom setup on the 10.12.1 Mac was around with older versions as far back as a year ago, while the one on the 10.10.5 Mac was setup only about 3 months ago.  A new setup on 10.12.1 behaves the same.

It would be nice if the default behaviour could become like after the  @lee-dohm fix with an update for all Atom installations.

**Comment by 50Wliu** [comment]
I'll inquire about getting this info a 1.12 hotfix.  Otherwise, this week be fixed in Atom 1.14.0.

**Comment by TheCGCy** [comment]
The issue still there. Although the workaround works, just wished it'll be fix… officially.

**Comment by Ingramz** [comment]
Could you please provide Atom version you are using and the file that is
not being opened with expected highlighting?

R, 13. jaanuar 2017 kell 18:08 Cy <notifications@github.com> kirjutas:

**Comment by TheCGCy** [comment]
My current OS version are 10.11.6, this issue appears on both Atom 1.12.9 and 1.13.0. the file are any MacOS *plist file. 

For example the plist file within atom editor itself. 
[Info.plist.zip](https://github.com/atom/language-property-list/files/705139/Info.plist.zip)
![screen shot 2017-01-14 at 3 22 18 am](https://cloud.githubusercontent.com/assets/10301646/21942603/b79f8ecc-da08-11e6-9da8-7b041d31e358.png)

**Comment by 50Wliu** [comment]
Like I said in my comment, this will only be fixed starting with Atom 1.14.0, currently in beta.

**Comment by erikng** [comment]
I've been using the Atom beta's but with 1.14.1 now official I went back to it.

Plists are opening as old style. Did it not make the official release?
